### PR TITLE
fix(examples) Fix Opacus in Flower example

### PR DESCRIPTION
**Pull request by mohammadnaseri**
## Proposal
This example adapts the existing Opacus example, incorporating local differential privacy with sample-level privacy using the  [Opacus](https://opacus.ai/) privacy engine with PyTorch.

### CODE REVIEW

File: examples/opacus/pyproject.toml [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
+description = "Sample Differential Privacy with Opacus in Flower"
+authors = [
+    { name = "The Flower Authors", email = "hello@flower.ai" },
+]
+dependencies = [
+    "flwr>=1.8.0,<2.0",
+    "flwr-datasets[vision]>=0.0.2,<1.0.0",
+    "torch==2.1.1",
+    "torchvision==0.16.1",
+    "tqdm==4.65.0"
```
Comment (jafermarq) [inline examples/opacus/pyproject.toml:17]:
```suggestion
    "tqdm==4.65.0",
```

File: examples/opacus/pyproject.toml [hunk 2]
```diff
@@ -1,15 +1,22 @@
 [build-system]
-requires = ["poetry-core>=1.4.0"]
-build-backend = "poetry.core.masonry.api"
+requires = ["hatchling"]
+build-backend = "hatchling.build"
 
-[tool.poetry]
-name = "flwr_opacus"
+[project]
+name = "opacus"
```
Comment (jafermarq) [inline examples/opacus/pyproject.toml:6]:
```suggestion
name = "opacus-fl"
```

Comment (jafermarq) [reply examples/opacus/pyproject.toml:6]: it seems it cannot be the same name as the `opacus` package

File: examples/opacus/README.md [hunk 3]
```diff
@@ -1,28 +1,54 @@
...
+```bash
+flower-client-app client:app1 --insecure
+```
+
+### 3. Run the Flower App
+
+With both the long-running server (SuperLink) and two clients (SuperNode) up and running, we can now run the actual Flower App:
+
+```bash
+flower-server-app server:app2 --insecure
```
Comment (jafermarq) [inline examples/opacus/README.md:53]:
```suggestion
flower-server-app server:app --insecure
```

File: examples/opacus/README.md [hunk 4]
```diff
@@ -1,28 +1,54 @@
...
 
-Run the server with `python server.py`. Then open two (or more) new terminals to start two (or more) clients with `python dp_cifar_client.py`.
+### Installing dependencies
 
-## Running the simulation
+Project dependencies are defined in `requirements.txt`. Install them with:
 
-Note: It is not possible to see the total privacy budget used with this example since the simulation creates clients from scratch every round.
+```shell
+pip install -r requirements.txt
```
Comment (jafermarq) [inline examples/opacus/README.md:29]:
```suggestion
pip install .
```

File: examples/opacus/client.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        train_loader, test_loader = load_data(partition_id=partition_id)
+        return FlowerClient(
+            net,
+            train_loader,
+            test_loader,
+            target_delta,
+            noise_multiplier,
+            max_grad_norm,
+        ).to_client()
+
```
Comment (jafermarq) [inline examples/opacus/client.py:161]:
```suggestion
        return client_fn
```

File: examples/opacus/README.md [hunk 6]
```diff
@@ -1,28 +1,60 @@
-# Differentially Private Federated Learning using Opacus, PyTorch and Flower
+# Training with Sample-Level Differential Privacy using Opacus Privacy Engine
 
-This example contains code demonstrating how to include the Opacus library for training a model using DP-SGD. The code is adapted from multiple other examples:
+In this example, we demonstrate how to train a model with differential privacy (DP) using Flower. We employ PyTorch and integrate the Opacus Privacy Engine to achieve sample-level differential privacy. This setup ensures robust privacy guarantees during the client training phase. The code is adapted from the [PyTorch Quickstart example](https://github.com/adap/flower/tree/main/examples/quickstart-pytorch).
 
-- PyTorch Quickstart
-- Simulation Quickstart
-- Simulation Extended Example
+For more information about DP in Flower please refer to the [tutorial](https://flower.ai/docs/framework/how-to-use-differential-privacy.html). For additional information about Opacus, visit the official \[website\] (https://opacus.ai/).
```
Comment (jafermarq) [inline examples/opacus/README.md:5]:
```suggestion
For more information about DP in Flower please refer to the [tutorial](https://flower.ai/docs/framework/how-to-use-differential-privacy.html). For additional information about Opacus, visit the official [website](https://opacus.ai/).
```

File: examples/opacus/README.md [hunk 7]
```diff
@@ -1,28 +1,60 @@
...
+```shell
+git clone --depth=1 https://github.com/adap/flower.git && mv flower/examples/sample-level-dp-opacus . && rm -rf flower && cd sample-level-dp-opacus
+```
 
-The parameters can be set in `dp_cifar_main.py`.
+This will create a new directory called `sample-level-dp-opacus` containing the following files:
 
-## Running the client
+```shell
+-- requirements.txt
```
Comment (jafermarq) [inline examples/opacus/README.md:18]:
```suggestion
-- pyproject.toml
```

File: examples/opacus/README.md [hunk 8]
```diff
@@ -1,28 +1,60 @@
...
+-- client.py
+-- server.py
+-- README.md
+```
 
-Run the server with `python server.py`. Then open two (or more) new terminals to start two (or more) clients with `python dp_cifar_client.py`.
+### Installing dependencies
 
-## Running the simulation
+Project dependencies are defined in `requirements.txt`. Install them with:
```
Comment (jafermarq) [inline examples/opacus/README.md:26]:
```suggestion
Project dependencies are defined in `pyproject.toml`. Install them with:
```

File: examples/opacus/README.md [hunk 9]
```diff
@@ -1,28 +1,60 @@
...
 
-- **Flower** nightly release (or development version from `main` branch) for the simulation, otherwise normal Flower for the client
-- **PyTorch** 1.7.1 (but most likely will work with older versions)
-- **Ray** 1.4.1 (just for the simulation)
-- **Opacus** 0.14.0
+Start by cloning the example. We prepared a single-line command that you can copy into your shell which will checkout the example for you:
 
-## Privacy Parameters
+```shell
+git clone --depth=1 https://github.com/adap/flower.git && mv flower/examples/sample-level-dp-opacus . && rm -rf flower && cd sample-level-dp-opacus
```
Comment (jafermarq) [inline examples/opacus/README.md:12]: This will only work if we rename the directory of this example as `sample-level-dp-opacus`

File: examples/opacus/client.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+            correct += (torch.max(outputs.data, 1)[1] == labels).sum().item()
+    accuracy = correct / len(test_loader.dataset)
+    return loss, accuracy
+
+
+def load_data(partition_id):
+    fds = FederatedDataset(dataset="cifar10", partitioners={"train": 2})
+    partition = fds.load_partition(partition_id)
+    # Divide data on each node: 80% train, 20% test
+    partition_train_test = partition.train_test_split(test_size=0.2)
```
Comment (jafermarq) [inline examples/opacus/client.py:74]:
```suggestion
    partition_train_test = partition.train_test_split(test_size=0.2, seed=42)
```

**Review by jafermarq — APPROVED** [review]
